ad400x-fmcz/ad400x.c

```diff
@@ -0,0 +1,190 @@
+ *  - Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  - Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Analog Devices, Inc. nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include "stdio.h"
+#include "stdlib.h"
+#include "stdbool.h"
+#include "ad400x.h"
+#include "spi_engine.h"
+
+/******************************************************************************/
+/************************** Functions Implementation **************************/
+/******************************************************************************/
+
+/**
+ * Read from device.
+ * @param dev - The device structure.
+ * @param reg_data - The register data.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad400x_spi_reg_read(struct ad400x_dev *dev,
+			    uint8_t *reg_data)
+{
+	int32_t ret;
+	uint8_t buf[2];
+	uint32_t spi_clk_hz_tmp;
+
+	buf[0] = AD400X_READ_COMMAND;
+	buf[1] = 0xFF;
+
+	// register access runs at a lower clock rate (~2MHz)
+	spi_clk_hz_tmp = dev->spi_desc->spi_clk_hz;
+	dev->spi_desc->spi_clk_hz = 2000000;
```

Comment:
maybe make this a parameter somewhere ; maybe `dev->spi_desc` ; something like `spi_clk_hz_reg_access` or similar
2 Mhz may be too fast for some setups ; and it's not a good idea to assume SPI speed in general;

so, then what could be done is:
```
uint32_t spi_clk_hz_tmp = (uint32_t) -1;

if (dev->spi_desc->spi_clk_hz > <spi_clk_hz_reg_access> ) {
    spi_clk_hz_tmp = dev->spi_desc->spi_clk_hz;
    dev->spi_desc->spi_clk_hz = <spi_clk_hz_reg_access>;
}
.............

if (spi_clk_hz_tmp != (uint32_t) -1)
    dev->spi_desc->spi_clk_hz = spi_clk_hz_tmp;
```